File: apps/python-api/app/store/victoriametrics.py
from datetime import datetime, timedelta, timezone

from app.models import Sample, Tag
from app.store.catalog import CatalogMem
from app.store.net import _utc, ilp_float, new_client, raise_status, unix_ns

_HTTP = None


def _http():
    global _HTTP
    if _HTTP is None or _HTTP.is_closed:
        _HTTP = new_client(timeout=15.0)
    return _HTTP


_LOOKBACK = timedelta(hours=2)  # archive max gap is 1h; 370d last_over_time is overkill


def _tag_re(tag_ids: list[int]) -> str:
    return "|".join(str(i) for i in tag_ids)


class VictoriaMetricsStore:
    name = "victoriametrics"

    def __init__(self, url: str) -> None:
        self._url = url.rstrip("/")
        self._tags = CatalogMem()

    async def ping(self) -> None:
        resp = await _http().get(f"{self._url}/health")
        raise_status(resp, "vm health")

    async def write(self, samples: list[Sample]) -> None:
        if not samples:
            return
        # Same ILP as Go: empty measurement, quality label, field prism_sample.
        parts = [
            f",tag_id={s.tag_id},quality={s.quality} prism_sample={ilp_float(s.value)} {unix_ns(s.ts)}\n"
            for s in samples
        ]
        resp = await _http().post(
            f"{self._url}/write",
            params={"precision": "ns"},
            content="".join(parts).encode("ascii"),
            headers={"Content-Type": "text/plain"},
        )
        raise_status(resp, "vm write")

    async def locf(self, tag_ids: list[int], at: datetime) -> list[Sample]:
        seed, _ = await self._scan(tag_ids, at - _LOOKBACK, at, at, at, with_mid=False)
        return seed

    async def range(self, tag_ids: list[int], start: datetime, end: datetime) -> list[Sample]:
        seed, mid = await self._scan(tag_ids, start - _LOOKBACK, end, start, end, with_mid=True)
        return seed + mid

    async def _scan(
        self,
        tag_ids: list[int],
        export_start: datetime,
        export_end: datetime,
        from_: datetime,
        to: datetime,
        with_mid: bool,
    ) -> tuple[list[Sample], list[Sample]]:
        if not tag_ids:
            return [], []
        start = _utc(from_)
        stop = _utc(to)
        params = {
            "match[]": f'prism_sample{{tag_id=~"{_tag_re(tag_ids)}"}}',
            "start": int(_utc(export_start).timestamp()),
            "end": int(_utc(export_end).timestamp()),
            "format": "tag_id,quality,__value__,__timestamp__:unix_ms",
        }
        best: dict[int, Sample] = {}
        mid: list[Sample] = []
        async with _http().stream("GET", f"{self._url}/api/v1/export/csv", params=params) as resp:
            if resp.status_code >= 300:
                body = (await resp.aread())[:500]
                raise RuntimeError(f"vm export {resp.status_code}: {body!r}")
            async for line in resp.aiter_lines():
                if not line or line.startswith("tag_id"):
                    continue
                row = line.split(",")
                if len(row) < 4:
                    continue
                try:
                    tag_id = int(row[0])
                except ValueError:
                    continue
                quality = int(float(row[1] or 0))
                val = float(row[2] or 0)
                when = datetime.fromtimestamp(int(row[3]) / 1000, tz=timezone.utc)
                if when <= start:
                    prev = best.get(tag_id)
                    if prev is None or when > prev.ts:
                        best[tag_id] = Sample(
                            ts=when, tag_id=tag_id, value=val, quality=quality, carried=with_mid
                        )
                    continue
                if with_mid and when <= stop:
                    mid.append(Sample(ts=when, tag_id=tag_id, value=val, quality=quality))
        seed = [best[tag_id] for tag_id in tag_ids if tag_id in best]
        return seed, mid

    async def upsert_tags(self, tags: list[Tag]) -> None:
        self._tags.upsert(tags)

    async def list_tags(self) -> list[Tag]:
        return self._tags.list()

    async def close(self) -> None:
        global _HTTP
        if _HTTP is not None and not _HTTP.is_closed:
            await _HTTP.aclose()
        _HTTP = None

File: apps/go-api/internal/store/victoriametrics.go
package store

import (
	"bytes"
	"context"
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"prism/go-api/internal/model"
)

type VictoriaMetrics struct {
	base     string
	writeURL string
	client   *http.Client
	tags     catalogMem
}

func NewVictoriaMetrics(base string) (*VictoriaMetrics, error) {
	base = strings.TrimRight(base, "/")
	return &VictoriaMetrics{
		base:     base,
		writeURL: base + "/write?precision=ns",
		client:   newWriteHTTPClient(15 * time.Second),
	}, nil
}

func (s *VictoriaMetrics) Name() string { return "victoriametrics" }

func (s *VictoriaMetrics) Close() error {
	s.client.CloseIdleConnections()
	return nil
}

func (s *VictoriaMetrics) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.base+"/health", nil)
	if err != nil {
		return err
	}
	resp, err := s.client.Do(req)
	if err != nil {
		return err
	}
	defer closeHTTP(resp)
	if resp.StatusCode >= 300 {
		return fmt.Errorf("vm health status %d", resp.StatusCode)
	}
	return nil
}

func (s *VictoriaMetrics) Write(ctx context.Context, samples []model.Sample) error {
	if len(samples) == 0 {
		return nil
	}
	buf := getBuf()
	defer putBuf(buf)
	for i := range samples {
		p := &samples[i]
		// Canonical ILP for all APIs: empty measurement, quality is a label,
		// field name prism_sample is the metric. Query match[] is prism_sample{tag_id}.
		buf.WriteString(",tag_id=")
		buf.WriteString(strconv.FormatUint(uint64(p.TagID), 10))
		buf.WriteString(",quality=")
		buf.WriteString(strconv.FormatUint(uint64(p.Quality), 10))
		buf.WriteString(" prism_sample=")
		buf.WriteString(strconv.FormatFloat(p.Value, 'g', -1, 64))
		buf.WriteByte(' ')
		buf.WriteString(strconv.FormatInt(p.TS.UTC().UnixNano(), 10))
		buf.WriteByte('\n')
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.writeURL, bytes.NewReader(buf.Bytes()))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "text/plain")
	resp, err := s.client.Do(req)
	if err != nil {
		return err
	}
	defer closeHTTP(resp)
	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return fmt.Errorf("vm write status %d: %s", resp.StatusCode, body)
	}
	return nil
}

func (s *VictoriaMetrics) Locf(ctx context.Context, tagIDs []uint32, at time.Time) ([]model.Sample, error) {
	at = at.UTC()
	seed, _, err := s.scanExport(ctx, tagIDs, at.Add(-vmLookback), at, at, at, false)
	return seed, err
}

func (s *VictoriaMetrics) Range(ctx context.Context, tagIDs []uint32, from, to time.Time) ([]model.Sample, error) {
	from, to = from.UTC(), to.UTC()
	seed, mid, err := s.scanExport(ctx, tagIDs, from.Add(-vmLookback), to, from, to, true)
	if err != nil {
		return nil, err
	}
	return append(seed, mid...), nil
}

// scanExport pulls raw samples in one /api/v1/export/csv call.
// Archive max gap is 1h (hourly tags); 2h lookback is enough for locf at 364d ago.
func (s *VictoriaMetrics) scanExport(ctx context.Context, tagIDs []uint32, start, end, from, to time.Time, withMid bool) ([]model.Sample, []model.Sample, error) {
	if len(tagIDs) == 0 {
		return nil, nil, nil
	}
	params := url.Values{}
	params.Set("match[]", fmt.Sprintf(`prism_sample{tag_id=~"%s"}`, vmTagRE(tagIDs)))
	params.Set("start", strconv.FormatInt(start.Unix(), 10))
	params.Set("end", strconv.FormatInt(end.Unix(), 10))
	params.Set("format", "tag_id,quality,__value__,__timestamp__:unix_ms")
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.base+"/api/v1/export/csv?"+params.Encode(), nil)
	if err != nil {
		return nil, nil, err
	}
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, nil, err
	}
	defer closeHTTP(resp)
	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 2048))
		return nil, nil, fmt.Errorf("vm export status %d: %s", resp.StatusCode, body)
	}

	best := make(map[uint32]model.Sample, len(tagIDs))
	var mid []model.Sample
	cr := csv.NewReader(resp.Body)
	cr.ReuseRecord = true
	cr.FieldsPerRecord = -1
	for {
		row, err := cr.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, nil, err
		}
		if len(row) < 4 {
			continue
		}
		if row[0] == "tag_id" {
			continue
		}
		id64, err := strconv.ParseUint(row[0], 10, 32)
		if err != nil {
			continue
		}
		id := uint32(id64)
		q, _ := strconv.Atoi(row[1])
		val, _ := strconv.ParseFloat(row[2], 64)
		ms, _ := strconv.ParseInt(row[3], 10, 64)
		t := time.UnixMilli(ms).UTC()
		if !t.After(from) {
			if prev, ok := best[id]; !ok || t.After(prev.TS) {
				best[id] = model.Sample{TS: t, TagID: id, Value: val, Quality: uint16(q), Carried: withMid}
			}
			continue
		}
		if withMid && !t.After(to) {
			mid = append(mid, model.Sample{TS: t, TagID: id, Value: val, Quality: uint16(q)})
		}
	}

	seed := make([]model.Sample, 0, len(tagIDs))
	for _, id := range tagIDs {
		if sample, ok := best[id]; ok {
			seed = append(seed, sample)
		}
	}
	return seed, mid, nil
}

func (s *VictoriaMetrics) UpsertTags(_ context.Context, tags []model.Tag) error {
	s.tags.upsert(tags)
	return nil
}

func (s *VictoriaMetrics) ListTags(_ context.Context) ([]model.Tag, error) {
	return s.tags.list(), nil
}

const vmLookback = 2 * time.Hour

func vmTagRE(ids []uint32) string {
	parts := make([]string, len(ids))
	for i, id := range ids {
		parts[i] = strconv.FormatUint(uint64(id), 10)
	}
	return strings.Join(parts, "|")
}
